src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c

```diff
@@ -208,88 +218,129 @@ static inline void reg_wr(void __iomem *reg, u32 val)
 }
 
 static int icap_cache_bitstream_axlf_section(struct platform_device *pdev,
-	const struct axlf *xclbin, enum axlf_section_kind kind);
+	const struct axlf *xclbin, enum axlf_section_kind kind, uint32_t slot_id);
 static void icap_set_data(struct icap *icap, struct xcl_pr_region *hwicap);
 static uint64_t icap_get_data_nolock(struct platform_device *pdev, enum data_kind kind);
 static uint64_t icap_get_data(struct platform_device *pdev, enum data_kind kind);
 static void icap_refresh_addrs(struct platform_device *pdev);
-static int icap_calib_and_check(struct platform_device *pdev);
+static int icap_calib_and_check(struct platform_device *pdev, uint32_t slot_id);
 static void icap_probe_urpdev(struct platform_device *pdev, struct axlf *xclbin,
-	int *num_urpdev, struct xocl_subdev **urpdevs);
+	int *num_urpdev, struct xocl_subdev **urpdevs, uint32_t slot_id);
+
+static struct islot_info *
+icap_slot_init(struct icap *icap, uint32_t slot_id)
+{
+	struct islot_info *islot = icap->slot_info[slot_id];
+	
+	if (islot) {
+		vfree(islot);
+	}
 
-static int icap_xclbin_wr_lock(struct icap *icap)
+	islot = vzalloc(sizeof(struct islot_info));
+	if (!islot) {
+		ICAP_ERR(icap, "Memory allocation failure");
+		return NULL;
+	}
+	
+	init_waitqueue_head(&islot->reader_wq);
+	islot->slot_idx = slot_id;
+
+	return islot;
+}
+
+static int icap_xclbin_wr_lock(struct icap *icap, uint32_t slot_id)
 {
+	struct islot_info *islot = icap->slot_info[slot_id];
 	pid_t pid = pid_nr(task_tgid(current));
 	int ret = 0;
 
+	if (!islot)
+		return -EINVAL;
+		
 	mutex_lock(&icap->icap_lock);
-	if (icap->busy) {
+	if (islot->busy) {
 		ret = -EBUSY;
 	} else {
-		icap->busy = (u64)pid;
+		islot->busy = (u64)pid;
 	}
 	mutex_unlock(&icap->icap_lock);
 
 	if (ret)
 		goto done;
 
-	ret = wait_event_interruptible(icap->reader_wq, icap->reader_ref == 0);
+	ret = wait_event_interruptible(islot->reader_wq,
+				       islot->reader_ref == 0);
 
 	if (ret)
 		goto done;
 
-	BUG_ON(icap->reader_ref != 0);
+	BUG_ON(islot->reader_ref != 0);
```

Comment:
We should move "islot = icap->slot_info[slot_id];" into the critical section. The icap_lock is also protecting slot_info array.

Reply:
I am not sure, As we fallback to icap lock, do we need to protect the slot array or not. 
This is the same behavior as of today without slot changes. Am i missing something ?

Reply:
the icap lock was used to protect icap data structure, now you added slot array to it, why it's not protected by it?

Reply:
Done. Please check. Thanks